sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEUtils.scala

```diff
@@ -57,4 +58,22 @@ object AQEUtils {
       }
     case _ => Some(UnspecifiedDistribution)
   }
+
+  // Add an extra shuffle if input plan does not satisfy the required distribution.
+  // This method is invoked after optimizing skewed join in case we change final stage
+  // output partitioning.
+  def ensureRequiredDistribution(
```

Comment:
Can we take advantage of `EnsureRequirements` here by passing it a root distro requirement instead of using this adhoc logic?

Reply:
good point, merged this code into `EnsureRequirements`